x-pack/plugins/ml/server/models/data_frame_analytics/analytics_manager.ts

```diff
@@ -0,0 +1,437 @@
+        // fetch transform so we can get original index pattern
+        const transformData = await this.getTransformData(id);
+        return { transformData, isTransform: true };
+      }
+    } catch (error) {
+      throw Boom.badData(error.message ? error.message : error);
+    }
+  }
+
+  private getAnalyticsModelElements(
+    analyticsId: string
+  ): {
+    modelElement?: AnalyticsMapNodeElement;
+    modelDetails?: any;
+    edgeElement?: AnalyticsMapEdgeElement;
+  } {
+    // Get inference model for analytics job and create model node
+    const analyticsModel = this.findJobModel(analyticsId);
+    let modelElement;
+    let edgeElement;
+
+    if (analyticsModel !== undefined) {
+      const modelId = `${analyticsModel.model_id}-${JOB_MAP_NODE_TYPES.INFERENCE_MODEL}`;
+      modelElement = {
+        data: {
+          id: modelId,
+          label: analyticsModel.model_id,
+          type: JOB_MAP_NODE_TYPES.INFERENCE_MODEL,
+        },
+      };
+      // Create edge for job and corresponding model
+      edgeElement = {
+        data: {
+          id: `${analyticsId}-${JOB_MAP_NODE_TYPES.ANALYTICS}~${modelId}`,
+          source: `${analyticsId}-${JOB_MAP_NODE_TYPES.ANALYTICS}`,
+          target: modelId,
+        },
+      };
+    }
+
+    return { modelElement, modelDetails: analyticsModel, edgeElement };
+  }
+
+  /**
+   * Works backward from jobId to return related jobs from source indices
+   * @param jobId
+   */
+  async getAnalyticsMap(analyticsId: string): Promise<AnalyticsMapReturnType> {
+    const result: any = { elements: [], details: {}, error: null };
+    const modelElements: MapElements[] = [];
+
+    try {
+      await this.setInferenceModels();
+      // Create first node for incoming analyticsId
+      let data = await this.getAnalyticsJobData(analyticsId);
+      let nextLinkId = data?.source?.index[0];
+      let nextType = JOB_MAP_NODE_TYPES.INDEX_PATTERN;
+      let complete = false;
+      let link: NextLinkReturnType;
+      let count = 0;
+      let rootTransform;
+      let rootIndexPattern;
+
+      const firstNodeId = `${data.id}-${JOB_MAP_NODE_TYPES.ANALYTICS}`;
+
+      result.elements.push({
+        data: {
+          id: firstNodeId,
+          label: data.id,
+          type: JOB_MAP_NODE_TYPES.ANALYTICS,
+          analysisType: getAnalysisType(data?.analysis),
+        },
+      });
+      result.details[firstNodeId] = data;
+
+      let { modelElement, modelDetails, edgeElement } = this.getAnalyticsModelElements(analyticsId);
+      if (isAnalyticsMapNodeElement(modelElement)) {
+        modelElements.push(modelElement);
+        result.details[modelElement.data.id] = modelDetails;
+      }
+      if (isAnalyticsMapEdgeElement(edgeElement)) {
+        modelElements.push(edgeElement);
+      }
+      // Add a safeguard against infinite loops.
+      while (complete === false) {
+        count++;
+        if (count >= 100) {
+          break;
+        }
```

Comment:
For the fixed amount of iterations, a `for` loop is used

Reply:
The problem is we don't know how many iteration we need since we're working backwards from the job/model id. We need to keep going until we've hit the source index/transform. The count is just to protect against infinite loops in case something changes or a new case is introduced that we haven't accounted for yet.

Reply:
`for` loop also supports the `break` statement.